File: src/fetch.nim
import osproc
import strutils

import format_article

proc fetchLatest*(): int =
  var p = startProcess(
    "curl",
    args = [
      "-s", "-L", "-o", "/tmp/news-reader-latest",
      "https://github.com/redroselinux/news-reader/" &
      "raw/refs/heads/master/news/latest"
    ],
    options = {poUsePath}
  )

  if p.waitForExit() != 0:
    p.close()
    echo "\e[91m*\e[0m Failed to fetch latest news"
    return 1
  p.close()

  var latest: string
  try:
    latest = readFile("/tmp/news-reader-latest").strip()
  except:
    echo "\e[91m*\e[0m Failed to read latest news"
    return 1

  # latest is simply a url; fetch the url we just got
  p = startProcess(
    "curl",
    args = ["-s", "-L", "-o", "/tmp/news-reader-latest", latest],
    options = {poUsePath}
  )

  if p.waitForExit() != 0:
    p.close()
    echo "\e[91m*\e[0m Failed to fetch latest news"
    return 1
  p.close()

  try:
    latest = readFile("/tmp/news-reader-latest")
  except:
    echo "\e[91m*\e[0m Failed to read latest news"
    return 1
  if latest.strip.startsWith("<!DOCTYPE html>"):
    echo "\e[91m*\e[0m Failed to fetch latest"
    return 1
  
  echo formatArticle(latest)
  return 0

proc fetchSpecific*(date: string): int =
  var p = startProcess(
    "curl",
    args = [
      "-s", "-L", "-o", "/tmp/news-reader-latest",
      "https://github.com/redroselinux/news-reader/" &
      "raw/refs/heads/master/news/" & date
    ],
    options = {poUsePath}
  )

  if p.waitForExit() != 0:
    p.close()
    echo "\e[91m*\e[0m Failed to fetch article " & date
    return 1
  p.close()

  var latest: string
  try:
    latest = readFile("/tmp/news-reader-latest")
  except:
    echo "\e[91m*\e[0m Failed to read " & date
    return 1
  if latest.strip.startsWith("<!DOCTYPE html>"):
    echo "\e[91m*\e[0m Failed to fetch " & date
    return 1

  echo formatArticle(latest)
  return 0

File: src/fetch.c
#include "lib.h"
#include <string.h>

int read_latest() {
    // first fetch latest url, read from there and fetch that
    char *url = fetch(
        "https://github.com/redroselinux/news-reader/raw/refs/heads/master/news/latest"
    );
    if (strcmp("error", url) == 0) {
        puts("\e[1;91mx\e[0m Error during fetch");
        return 1;
    }

    // if there is a newline remove it
    if (url[strlen(url) - 1] == '\n') {
        url[strlen(url) - 1] = '\0';
    }

    // fetch the url we just got
    printf("\e[94m→\e[0m Fetching %s\n", url);
    char *content = fetch(url);
    free(url);
    if (strcmp("error", content) == 0) {
        puts("\e[1;91mx\e[0m Error during fetch");
        return 1;
    }

    view_pretty_article(content);

    free(content);
    return 0;
}

int read_article(char* article) {
    char url[512];
    snprintf(url, 512, "https://raw.githubusercontent.com/redroselinux/news-reader/refs/heads/master/news/%s", article);

    printf("\e[1;94m→\e[0m Fetching %s\n", url);
    char *content = fetch(url);
    if (strcmp("error", content) == 0) {
        puts("\e[1;91mx\e[0m Error during fetch");
        return 1;
    }
    if (!strcmp("404: Not Found", content)) {
        puts("\e[1;91mx\e[0m Article does not exist");
        return 1;
    }

    view_pretty_article(content);

    free(content);
    return 0;
}
